File: Graph Theory Peaks/BFS/Q1. Find if Path Exists in Graph/Q1_FindifPathExistsinGraph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        
        // Create adjacency list representation of graph
        Map<Integer, List<Integer>> graph = new HashMap<>();
        
        // Build undirected graph
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            
            // Add v to u's list
            graph.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
            
            // Add u to v's list (because graph is undirected)
            graph.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
        }
        
        // Set to track visited nodes (to avoid cycles)
        Set<Integer> visited = new HashSet<>();
        
        // Start DFS from source
        return dfs(source, destination, graph, visited);
    }
    
    private boolean dfs(int node, int destination, Map<Integer, List<Integer>> graph, Set<Integer> visited) {
        
        // If we reached destination → path exists
        if (node == destination) {
            return true;
        }
        
        // Mark current node as visited
        visited.add(node);
        
        // Traverse all neighbors of current node
        for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
            
            // Visit only unvisited neighbors
            if (!visited.contains(neighbor)) {
                
                // If any path leads to destination → return true
                if (dfs(neighbor, destination, graph, visited)) {
                    return true;
                }
            }
        }
        
        // If no path found
        return false;
    }
}

File: Graph Theory Peaks/BFS/Q1. Find if Path Exists in Graph/Q1_FindifPathExistsinGraph.py
import collections


class Solution(object):
    def validPath(self, n, edges, source, destination):
        """
        :type n: int
        :type edges: List[List[int]]
        :type source: int
        :type destination: int
        :rtype: bool
        """
        
        # Create adjacency list using dictionary
        graph = collections.defaultdict(list)
        
        # Build undirected graph
        for u, v in edges:
            graph[u].append(v)  # add edge u -> v
            graph[v].append(u)  # add edge v -> u
        
        # DFS function to check if path exists
        def dfs(node, visited):
            
            # If we reach destination, path exists
            if node == destination:
                return True
            
            # Mark current node as visited
            visited.add(node)
            
            # Explore all neighbors
            for neighbor in graph[node]:
                
                # Visit only unvisited nodes
                if neighbor not in visited:
                    
                    # If any recursive call reaches destination → return True
                    if dfs(neighbor, visited):
                        return True
            
            # If no path found
            return False
        
        # Set to track visited nodes (avoid infinite loops)
        visited = set()
        
        # Start DFS from source
        return dfs(source, visited)